upload-api/tables/index.js

```diff
@@ -76,4 +76,20 @@ export const consumerTableProps = {
     consumer: { partitionKey: 'consumer', projection: ['provider', 'subscription'] },
     provider: { partitionKey: 'provider', projection: ['consumer'] }
   }
+}
+
+/** @type TableProps */
+export const rateLimitTableProps = {
+  fields: {
+    id: 'string',           // arbitrary identifier for this limit
+    cause: 'string',        // `baf...x` (CID of invocation that created record)
+    subject: 'string',      // string (arbitrary string identifying the subject to be limited)
+    rate: 'number',         // unitless number representing the rate to which the subject is limited
+    insertedAt: 'string',   // `2022-12-24T...`
+    updatedAt: 'string',    // `2022-12-24T...`
+  },
+  primaryIndex: { partitionKey: 'id' },
+  globalIndexes: {
+    subject: { partitionKey: 'subject', projection: ['rate', 'id'] },
+  }
```

Comment:
if folks will mostly query by `subject` then consider modelling the table so that the subject is the partition (primary) key, so we can get back all rate limits for that subject without a secondary index. Something like

```suggestion
export const rateLimitTableProps = {
  fields: {
    id: 'string',           // arbitrary identifier for this limit
    cause: 'string',        // `baf...x` (CID of invocation that created record)
    subject: 'string',      // string (arbitrary string identifying the subject to be limited)
    rate: 'number',         // unitless number representing the rate to which the subject is limited
    insertedAt: 'string',   // `2022-12-24T...`
    updatedAt: 'string',    // `2022-12-24T...`
  },
  primaryIndex: { partitionKey: 'subject', sortKey: 'id' }
```
if the `id` includes a timestamp such that lexigraphical sorting preserves rough ordering, then great!

The general recommendation with dynamo is to abuse the heck out of the primary key in ways you'd never consider if you've ever modelled data in anything else. It's wild.

> well-optimized single-table DynamoDB layout looks more like machine code than a simple spreadsheet
- https://www.trek10.com/blog/dynamodb-single-table-relational-modeling

(we're not ready for single table design yet, so this is a more advanced walk through then is strictly relevant but i think i shows the point...)

Reply:
mmm, that's a good thought... given your comments in the spec repo and a general desire to get this out I'm going to stick with the current model, but I think this would be a good idea to consider when we iterate on this - the migration should be straightforward when we get there!